config/rbac/role.yaml

```diff
@@ -6,6 +6,17 @@ metadata:
   creationTimestamp: null
   name: external-dns-operator
 rules:
+- apiGroups:
+  - ""
+  resources:
+  - endpoints
```

Comment:
Oouf, it took me some time to find out all about EndpointSlices and dualstack support.

1. This cluster role has the same resources as [the upstream one](https://github.com/kubernetes-sigs/external-dns/blob/master/kustomize/external-dns-clusterrole.yaml#L7).
2. `EndpointSlice` is not used anywhere in the code of ExternalDNS
3. No `AAAA` record is created for dualstack sources:
```bash
# ClusterIP source service
$ oc get svc httpserver2 -o yaml | grep -A 14 '^spec:'
spec:
  clusterIP: 172.30.46.13
  clusterIPs:
  - 172.30.46.13
  - fd02::9bfb
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    run: httpserver

# externalDNS logs    
$ oc -n external-dns logs deploy/external-dns-sample | grep 'publish-external-dns/httpserver2'
time="2021-09-16T15:00:43Z" level=debug msg="Endpoints generated from service: publish-external-dns/httpserver2: [httpserver2.external-dns-test.alebedev.com 0 IN A  172.30.46.13 []]"
time="2021-09-16T15:01:43Z" level=debug msg="Endpoints generated from service: publish-external-dns/httpserver2: [httpserver2.external-dns-test.alebedev.com 0 IN A  172.30.46.13 []]"

# A record which returns IPv4 ClusterIP
aws route53 list-resource-record-sets --output json --hosted-zone-id "/hostedzone/Z06076472WHIZX9L24NVR" --query "ResourceRecordSets[?Type == 'A']"
[
    {
        "Name": "httpserver2.external-dns-test.alebedev.com.",
        "Type": "A",
        "TTL": 300,
        "ResourceRecords": [
            {
                "Value": "172.30.46.13"
            }
        ]
    }
]
```

```bash
# LoadBalancer service source
$ oc get svc httpserver -o yaml | grep -A 25 '^spec:'
spec:
  clusterIP: 172.30.51.48
  clusterIPs:
  - 172.30.51.48
  - fd02::30c0
  externalTrafficPolicy: Cluster
  ipFamilies:
  - IPv4
  - IPv6
  ipFamilyPolicy: PreferDualStack
  ports:
  - nodePort: 31189
    port: 8080
    protocol: TCP
    targetPort: 8080
  selector:
    run: httpserver
  sessionAffinity: None
  type: LoadBalancer
status:
  loadBalancer:
    ingress:
    - hostname: abad68df9bcad4b7387d8a66d8fc2758-743607947.us-east-2.elb.amazonaws.com

# A record which redirects to AWS ELB DNS name (from Service's status)    
$ aws route53 list-resource-record-sets --output json --hosted-zone-id "/hostedzone/Z06076472WHIZX9L24NVR" --query "ResourceRecordSets[?Type == 'A']"
[
    {
        "Name": "httpserver.external-dns-test.alebedev.com.",
        "Type": "A",
        "AliasTarget": {
            "HostedZoneId": "Z3AADJGX6KTTL2",
            "DNSName": "abad68df9bcad4b7387d8a66d8fc2758-743607947.us-east-2.elb.amazonaws.com.",
            "EvaluateTargetHealth": true
        }
    },
]
```
4. For the moment the dualstack support is limited to the Ingress source which is backed-up by [ALB](https://github.com/kubernetes-sigs/external-dns/blob/43d29f8032d2e79d2d666f794b89b71ad3349a12/docs/tutorials/alb-ingress.md#dualstack-albs) or [Zalando's AWS](https://github.com/kubernetes-sigs/external-dns/blob/7b8c69b22e33e04041c3a8299e13cab4f44ad486/docs/tutorials/kube-ingress-aws.md#dualstack-albs) ingress controller. For such ingress objects it's possible to add `alb.ingress.kubernetes.io/ip-address-type: dualstack` annotation which will instruct the externalDNS to create `AAAA` record which will target tthe IP from the status of the ingress resources. Note: the status is populated by the controller with IPv4, IPv6 addresses of the LB created in cloud.

So, I don't think ExternalDNS does anything with `EndpointSlice` resource because even for the real dualstack cases it doesn't go that deep in Kubernetes network stack. As a matter of fact I'm not even sure that ExternalDNS needs rights for the `Endpoints` resource :) but I'd rather keep it as it's used in the upstream.

Overall IPv6 support is not very consolidated - mostly left into consideration of the provider. There is this issue though which tries to Xref all the effort: [2051](https://github.com/kubernetes-sigs/external-dns/issues/2051).

Reply:
ExternalDNS seems to take what is in `Service.Spec.ClusterIP` (not in `ClusterIPs`) and even if it's IPv6 address it tries to take it but then fails at IP validation:
```yaml
$ oc get svc httpserver-ipv6 -o yaml | grep -A 10 '^spec:'
spec:
  clusterIP: fd02::b244
  clusterIPs:
  - fd02::b244
  ipFamilies:
  - IPv6
  ipFamilyPolicy: SingleStack
  ports:
  - port: 8080
    protocol: TCP
    targetPort: 8080
```
```
time="2021-09-17T16:03:58Z" level=debug msg="Endpoints generated from service: publish-external-dns/httpserver-ipv6: [httpserver-ipv6.extdnstest.alebedev.com 0 IN A  fd02::b244 []]"
...
time="2021-09-17T16:05:00Z" level=error msg="InvalidChangeBatch: [Invalid Resource Record: 'FATAL problem: ARRDATAIllegalIPv4Address (Value is not a valid IPv4 address) encountered with 'fd02::b244'']\n\tstatus code: 400, request id: c762120b-aae3-42c3-b9bb-99418f7861b1"
```

This was already reported by [this upstream issue](https://github.com/kubernetes-sigs/external-dns/issues/1812).

Reply:
Comment in Jira epic: [link](https://issues.redhat.com/browse/NE-303?focusedCommentId=19051901&page=com.atlassian.jira.plugin.system.issuetabpanels%3Acomment-tabpanel#comment-19051901).